File: android/app/src/main/java/app/farmsy/android/core/Tips.kt
package app.farmsy.android.core

import io.ktor.client.request.get
import io.ktor.client.statement.bodyAsText
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.Serializable

/// Grootmoeders tips, as GET /api/tips serves them: one card each on the
/// Discover tab, about a product, with its photograph and a list action.
/// The Android twin of iOS Tips.swift.
@Serializable
data class Tip(
    val slug: String,
    val kicker: LocalizedText,
    val title: LocalizedText,
    val body: LocalizedText,
    val ingredient: String? = null,
    val image: String = "",
)

@Serializable
private data class TipsPayload(val tips: List<Tip> = emptyList())

object Tips {
    private val _tips = MutableStateFlow<List<Tip>>(emptyList())
    val tips: StateFlow<List<Tip>> = _tips.asStateFlow()

    suspend fun loadIfNeeded() {
        if (_tips.value.isNotEmpty()) return
        val payload = runCatching {
            val resp = httpClient.get("${Backend.WEB_API}/tips")
            if (resp.status.value != 200) null
            else lenientJson.decodeFromString<TipsPayload>(resp.bodyAsText())
        }.getOrNull() ?: return
        _tips.value = payload.tips
    }
}

File: Farmsy/Core/Tips.swift
import Foundation
import Observation

/// Grootmoeders tips, as GET /api/tips serves them: one card each on the
/// Discover tab, about a product, with its photograph and a list action.
struct Tip: Decodable, Identifiable, Sendable {
    let slug: String
    let kicker: LocalizedText
    let title: LocalizedText
    let body: LocalizedText
    let ingredient: String?
    let image: String
    var id: String { slug }
}

@MainActor
@Observable
final class Tips {
    static let shared = Tips()
    private(set) var tips: [Tip] = []
    private var loading: Task<Void, Never>?
    private init() {}

    func loadIfNeeded() async {
        if !tips.isEmpty { return }
        if let loading { await loading.value; return }
        let task = Task { [weak self] in
            let url = Backend.webAPI.appending(path: "tips")
            guard let (data, resp) = try? await URLSession.shared.data(from: url),
                  (resp as? HTTPURLResponse)?.statusCode == 200,
                  let decoded = try? JSONDecoder().decode(Payload.self, from: data)
            else { return }
            await MainActor.run { self?.tips = decoded.tips }
        }
        loading = task
        await task.value
        loading = nil
    }

    private struct Payload: Decodable { let tips: [Tip] }
}
